### DISCUSSION TITLE
Error when using proguard

### DISCUSSION THREAD

**Discussion by Vikviita** [joffrey-bion/krossbow#543] [General]
Missing class java.net.http.HttpResponse (referenced from: org.hildan.krossbow.websocket.ktor.HandshakeFailureDetails org.hildan.krossbow.websocket.ktor.HandshakeExceptionsJvmKt.extractHandshakeFailureDetails(java.net.http.WebSocketHandshakeException))
Missing class java.net.http.WebSocketHandshakeException (referenced from: org.hildan.krossbow.websocket.ktor.HandshakeFailureDetails org.hildan.krossbow.websocket.ktor.HandshakeExceptionsJvmKt.extractHandshakeFailureDetails(java.lang.Exception) and 1 other context)

**Comment by joffrey-bion** [comment]
Did you confirm that you don't get this error when not using proguard? 

Are you on Android? Note that if you're trying to use the JDK11's `HttpClient`, this is not available on Android.

**Reply by Vikviita** [reply]
yes 
this error only when 
isMinifyEnabled = true
 isShrinkResources = true
 
 I am on android 
 I use 
 
 krossbowKtor ={group = "org.hildan.krossbow", name = "krossbow-websocket-ktor", version.ref = "krossbowKtor" }
stompKtor = {group = "org.hildan.krossbow", name = "krossbow-stomp-core" ,version.ref= "krossbowKtor" }

**Reply by Vikviita** [reply]
in documentation said 
that krossbowKtor ={group = "org.hildan.krossbow", name = "krossbow-websocket-ktor", version.ref = "krossbowKtor" }
does not use HttpResponce from jdk11 
why proguard thinks that library referenced on it 
?

**Reply by joffrey-bion** [reply]
Just for the record, it is possible that Ktor uses the JDK11 client if you use the `Java` engine (see [Ktor engines](https://ktor.io/docs/client-engines.html)). But here I'm assuming you're using the `Android` or `OkHttp` engine, right? This should not be the cause, because it's only broken when proguard is enabled.

What might be happening is that Proguard complains because it doesn't understand that [my reference](https://github.com/joffrey-bion/krossbow/blob/137f35d7ae073908af46f15f8a116a28f4782eca/krossbow-websocket-ktor/src/jvmMain/kotlin/org/hildan/krossbow/websocket/ktor/HandshakeExceptionsJvm.kt#L14) to `java.net.http.WebSocketHandshakeException` is actually safe (I made sure it worked fine). Is there a way to force it to accept it?

**Reply by Vikviita** [reply]
It will work if use 

-dontwarn java.net.http.HttpResponse -dontwarn java.net.http.WebSocketHandshakeException

In Proguard rules

**Reply by joffrey-bion** [reply]
Ok great! Good to know.

I'm not sure if there is anything I could do here. We need to extract information from these exceptions when they are thrown (and thus exist on the classpath). This code should never throw at runtime even when the exception class is not on the classpath.

Basically, usages of JDK11 stuff will only be used when using the Ktor `Java` (JDK11) engine.
